File: 056/cpp/main.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        std::sort(intervals.begin(), intervals.end(), [](vector<int> a, vector<int> b){ return (a[0] < b[0]) || (a[0] == b[0] && a[1] < b[1]); });
        vector<vector<int>> results;
        for (auto &b: intervals) {
            if (results.empty()) {
                results.push_back(b);
                continue;
            }
            auto a = results.back();
            if (a[1] < b[0]) {
                results.push_back(b);
            } else {
                results.pop_back();
                results.push_back({a[0], std::max(a[1], b[1])});
            }
        }
        return results;
    }
};

File: 056/rust/main.rs
use std::cmp;

impl Solution {
    pub fn merge(intervals: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let mut intervals = intervals;
        intervals.sort_by_key(|a| (a[0], a[1]));
        let mut r = Vec::new();
        for b in intervals {
            if r.is_empty() {
                r.push(b.clone());
            }
            let a = r.pop().unwrap();
            if a[1] < b[0] {
                r.push(a);
                r.push(b);
            } else {
                r.push(vec![a[0], cmp::max(a[1], b[1])]);
            }
        }
        return r;
    }
}
